keras_rs/src/utils/ranking_metrics_utils.py

```diff
@@ -0,0 +1,240 @@
+            ops.shape(scores),
+            mask=mask,
+            shuffle_ties=True,
+            seed=seed,
+        )
+        scores = ops.take_along_axis(scores, shuffled_indices, axis=1)
+
+    # Get top-k indices.
+    _, indices = ops.top_k(scores, k=k, sorted=True)
+
+    # If we shuffled our `scores` tensor, we need to get the correct indices
+    # by indexing into `shuffled_indices`.
+    if shuffled_indices is not None:
+        indices = ops.take_along_axis(shuffled_indices, indices, axis=1)
+
+    return [
+        ops.take_along_axis(tensor_to_sort, indices, axis=1)
+        for tensor_to_sort in tensors_to_sort
+    ]
+
+
+def get_list_weights(
+    weights: types.Tensor, relevance: types.Tensor
+) -> types.Tensor:
+    """Computes per-list weights from provided sample weights.
+
+    The per-list weights are computed as:
+    ```
+    per_list_weights = sum(weights * relevance) / sum(relevance).
+    ```
+
+    For a list with sum(relevance) = 0, we set a default weight as the following
+    average weight while all the lists with sum(weights) = 0 are ignored:
+
+    ```
+    sum(per_list_weights) / num(sum(relevance) != 0 && sum(weights) != 0)
+    ```
+
+    When all the lists have sum(relevance) == 0, we set the average weight to
+    1.0.
+
+    This computation takes care of the following cases:
+    - When all the weights are 1.0, the per-list weights will be 1.0
+      everywhere, even for lists without any relevant examples because
+      `sum(per_list_weights) ==  num(sum(relevance) != 0)`. This handles the
+      standard ranking metrics where the weights are all
+      1.0.
+    - When every list has a nonzero weight, the default weight is not used.
+      This handles the unbiased metrics well.
+    - For the mixture of the above 2 scenario, the weights for lists with
+      nonzero relevance and nonzero weights is proportional to
+
+      ```
+      per_list_weights / sum(per_list_weights) *
+      num(sum(relevance) != 0) / num(lists)
+      ```
+
+      The rest have weights 1.0 / num(lists).
+
+    Args:
+        weights:  The weights `Tensor` of shape [batch_size, list_size].
+        relevance:  The relevance `Tensor` of shape [batch_size, list_size].
+
+    Returns:
+        A tensor of shape [batch_size, 1], containing the per-list weights.
+    """
+    # TODO: Pretty nasty function. Check if we can simplify it at a later point.
+
+    # Calculate if the sum of weights per list is greater than 0.0.
+    nonzero_weights = ops.greater(ops.sum(weights, axis=1, keepdims=True), 0.0)
+    # Calculate the sum of relevance per list
+    per_list_relevance = ops.sum(relevance, axis=1, keepdims=True)
+
+    # Identify lists where both weights and relevance sums are non-zero.
+    nonzero_relevance_condition = ops.greater(per_list_relevance, 0.0)
+    nonzero_relevance = ops.where(
+        nonzero_weights,
+        ops.cast(nonzero_relevance_condition, "float32"),
+        ops.zeros_like(per_list_relevance),
```

Comment:
Is there a reason to do `zeros_like` rather than just `0.0`? It should give the same result. But is it for ragged tensors or something?

Reply:
Does it typecast to the right dtype if we use 0.0? How does it know from which tensor it should take the dtype?

Reply:
Oh you would still need to do `ops.cast(0, per_list_relevance.dtype)`. But hopefully it would use less memory.